base/applications/mspaint/history.cpp

```diff
@@ -144,6 +161,29 @@ void ImageModel::PushImageForUndo(HBITMAP hbm)
     NotifyImageChanged();
 }
 
+void ImageModel::PushImageForUndo(const RECT& rcPartial)
+{
+    ATLTRACE("%s: %d\n", __FUNCTION__, m_currInd);
+
+    IMAGE_PART& part = m_historyItems[m_currInd];
+    part.clear();
+    part.m_bPartial = TRUE;
+    part.m_rcPart = rcPartial;
+
+    HBITMAP hbmMaster = LockBitmap();
+    part.m_hbmImage = getSubImage(hbmMaster, rcPartial);
+    UnlockBitmap(hbmMaster);
+
+    m_currInd = (m_currInd + 1) % HISTORYSIZE; // Go next
+
+    if (m_undoSteps < HISTORYSIZE - 1)
+        m_undoSteps++;
+    m_redoSteps = 0;
+
+    g_imageSaved = FALSE;
+    NotifyImageChanged();
```

Comment:
I wonder whether you could use the technique where the overloaded version does some specific stuff ,then calls the base function. Because it seems that there's a lot of duplicated code here.

Reply:
Fixed by using newly-added `ImageModel::PushDone`.